woocommerce/woocommerce-ios
# [Order Creation - Manual Taxes] Wrong subtotal when price includes taxes

Closes: #11010 


## Description

With this PR we fixed a bug where the order subtotal and total are wrongly calculated when we add an item whose price includes taxes. This happens when the item quantity is more than one. (See https://github.com/woocommerce/woocommerce-ios/issues/11010 and video)

The reason behind that is that we sync the order item with the price*quantity as subtotal, but this is not correct. Instead, we have to pass the base subtotal * quantity, otherwise, taxes will be calculated on the gross amount (price that includes taxes). This is obviously wrong, taxes should be calculated on the net, that is, the price minus taxes.

To fix that we rely on the subtotal amount we get from the backend when we synchronize the order by adding a new item. It seems like the backend is calculating it properly when we create an order even if we pass the wrong subtotal, but in the following cases, when updating the order with a new quantity, it fails because of our wrong passed subtotal.

This solution is similar to what we have on [Android](https://github.com/woocommerce/woocommerce-android/blob/13b9e5dedfce0fe0ae1ff30c3861a7916b4f440d/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L116)

`val pricePreDiscount = if (quantity == 0f) BigDecimal.ZERO else subtotal / quantity.toBigDecimal()
`

Which is applied [here](https://github.com/woocommerce/woocommerce-android/blob/13b9e5dedfce0fe0ae1ff30c3861a7916b4f440d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/ItemOperationsExt.kt#L12):

`val newSubtotal = pricePreDiscount.multiply(newQuantity.toBigDecimal())`

## Side effects

Here we're relying on the item sub-total that we got remotely on the order. If we create an order with an item with quantity 2, change the price of that product, and update the order by increasing the quantity of that item, it will use the previous subtotal, instead of the new price. In the current App Store version, if we change the price and then update the order, the order gets modified with the new price. I see both approaches with pros and cons.

## Testing instructions


Ensure that you have the setting prices entered with tax to yes in wp-admin/admin.php?page=wc-settings&tab=tax Also ensure that you have Display prices in the shop and Display prices during cart and checkout settings to Including Tax

1. Go to orders
2. Tap on + to create a new order.
3. Add a taxable product
4. See that the tax is calculated on the price without taxes. E.g. A product base price (without taxes) is $8. If we have a tax of 25%, the added tax for an item will be $2, thus being the displaying price $10
5. Increase the quantity to 2. Bug: The taxes are calculated on the total, and not on the base price (without taxes) as it should be. It wrongly shows a tax of 5 = (10 * 2)/4, but should be a tax of 4 = (8 * 2)/4, as it shows after this PR


## Screenshots


### Before

https://github.com/woocommerce/woocommerce-ios/assets/1864060/9653cd00-f9a6-4d31-9dd5-e83d6e690e73

### After

https://github.com/woocommerce/woocommerce-ios/assets/1864060/653e61ae-a20e-4c86-ba98-0984d6849b47

---
